src/actions.ts

```diff
@@ -69,7 +69,75 @@ export const joinLines = (editor: Editor) => {
   editor.setSelection(endOfCurrentLine);
 };
 
-export const duplicateLine = (editor: Editor) => {
+export const moveLineUp = (editor: Editor) => {
+  const selections = editor.listSelections();
+  if (selections.length === 0) {
+    return;
+  }
+  const { from, to } = getSelectionBoundaries(selections[0]);
+  if (from.line === 0) {
+    return;
+  }
+  // If we're selecting the final newline, that doesn't count as also selecting
+  // the next line, even though the end position is technically on that line.
+  const endingLine = to.line > from.line && to.ch === 0 ? to.line - 1 : to.line;
+  const previousLineStart = getLineStartPos(from.line - 1);
+  const fromLineStart = getLineStartPos(from.line);
+  const toLineEnd = getLineEndPos(endingLine, editor);
+  const contentsOfPreviousLine = editor.getRange(previousLineStart, fromLineStart);
+  editor.transaction({
+    changes: [
+      {text: '', from: previousLineStart, to: fromLineStart},
+      {text: '\n' + contentsOfPreviousLine.trimEnd(), from: toLineEnd},
+    ],
+    selection: {
+      from: { line: from.line - 1, ch: from.ch },
+      to: { line: to.line - 1, ch: to.ch }
+    },
+  });
```

Comment:
Does this allow you to group changes to be applied all at once instead of doing `replaceRange` and `setSelection` separately?

I've been meaning to look into `editor.transaction` as well but the Obsidian API for it seems to differ from the [CodeMirror 6 API](https://codemirror.net/6/docs/ref/#state.Transaction)...

Reply:
Yes exactly. Before I had two separate `replaceRange` calls but then Cmd+Z was stepping through them individually. I tried to merge into one `replaceRange` call but that left the selection in the wrong state. Then I found `transaction`.

Reply:
Nice! `transaction` seems like it would be especially useful for multiple selections, which I'm working on right now.